### PR TITLE
bpo-32888: Improve exception message in ast.literal_eval

### PR DESCRIPTION
**Pull request by Rosuav**
When a non-literal is given to literal_eval, attempt to be more
helpful with the message, rather than calling it 'malformed'.

https://bugs.python.org/issue32888

### CODE REVIEW

**Comment by merwok** [comment]
This looks good, do you mind opening a bugs.python.org ticket and adding a news entry?

**Comment by Rosuav** [comment]
Rebased onto current master, which includes shifting where the actual change happens. NEWS entry added, issue number added.

File: Lib/ast.py
Comment (serhiy-storchaka) [inline Lib/ast.py:57]: In what cases this exception is raised? Is it covered by tests?

File: Misc/NEWS.d/next/Library/2018-02-21-04-39-49.bpo-32888.G5i41Q.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-02-21-04-39-49.bpo-32888.G5i41Q.rst:2]: Please add "Patch by..."

**Review by serhiy-storchaka — COMMENTED** [review]
Please add a special test in test_ast.

File: Lib/ast.py
Comment (Rosuav) [reply Lib/ast.py:57]: The easiest way to trigger it is to use something that isn't permitted in literal_eval, but is valid syntax. There's already one such test - proving that f-strings aren't literal_evallable - and I guess there could be more, but since this is a whitelist, the error is basically "anything else, raise". The only difference in this patch is that the message is different for "syntactically valid but not acceptable by literal_eval" and "syntactically invalid".

File: Misc/NEWS.d/next/Library/2018-02-21-04-39-49.bpo-32888.G5i41Q.rst
Comment (Rosuav) [reply Misc/NEWS.d/next/Library/2018-02-21-04-39-49.bpo-32888.G5i41Q.rst:2]: Should that be done by amending the current commit, or adding an additional commit?

File: Lib/ast.py
Comment (serhiy-storchaka) [reply Lib/ast.py:57]: I meant what input triggers the "malformed node or string" error now?

File: Misc/NEWS.d/next/Library/2018-02-21-04-39-49.bpo-32888.G5i41Q.rst
Comment (serhiy-storchaka) [reply Misc/NEWS.d/next/Library/2018-02-21-04-39-49.bpo-32888.G5i41Q.rst:2]: It is better for review to add a new commit. In any case all commits are squashed before merging into the repository.

Comment (Rosuav) [reply Misc/NEWS.d/next/Library/2018-02-21-04-39-49.bpo-32888.G5i41Q.rst:2]: :+1:

File: Lib/ast.py
Comment (Rosuav) [reply Lib/ast.py:57]: Calling literal_eval with an AST node that has non-AST children in it. I don't think it's possible (post-patch) to trigger this error by calling literal_eval with a string, but I can't say for sure, so I don't want to say just "malformed node".

File: Lib/test/test_ast.py
Comment (briancurtin) [inline Lib/test/test_ast.py:595]: A helpful thing to use in between lines 594 and 595 would be `with self.subTest(test=test, message=message):` and then indent the parts below it. That way each of these end up being tested independently rather than potentially having the items loop have a case fail and cause the rest of the cases to fail to execute. 

Check out https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests

**Review by briancurtin — CHANGES_REQUESTED** [review]
Overall this looks good, just one suggestion on the test implementation.

**Comment by csabella** [comment]
@Rosuav, please address the last change request by @briancurtin and also rebase to fix the conflict.  Thanks!

**Comment by Rosuav** [comment]
BPO cites another couple of issues, one of which is still open, as dependencies. And it's not really a high priority - it's nothing more than a nice-to-have. There are a lot of barriers to getting patches accepted and it's often not worth the effort unless you REALLY care about the patch.

**Comment by merwok** [comment]
(Note that you can rebase or do a regular merge, which is often easier!  All PRs are squash merged in the end @csabella)

**Comment by merwok** [comment]
@serhiy-storchaka Should this wait for bpo-32893 per your last message on the ticket or could it be merged soon? (when conflicts are resolved + suggestion from Brian)

**Comment by vstinner** [comment]
A different approach using ast.dump(): PR #16620.

**Comment by vstinner** [comment]
What's the status of this PR?

@Rosuav: Are you still working on this?

---

> A different approach using ast.dump(): PR #16620.

Oh, I just closed it since it could produce very long error message: https://bugs.python.org/issue38396#msg354132

**Comment by Rosuav** [comment]
No, I'm not. It was a nice-to-have but between the pushback (no core dev seemed to want it to happen) and the constant need to fix merge conflicts meant that I abandoned this proposal ages ago. If someone else wants to champion this fix, then sure, but I don't have the energy to push everything, and this one wasn't worth it.

**Comment by vstinner** [comment]
@isidentical: Are you interested to take over this change? I'm not the good candidate to review it. Maybe @pablogsal or @serhiy-storchaka.

**Comment by isidentical** [comment]
Sure.

**Comment by csabella** [comment]
Closing this pull request as it is now GH-17662.

**Pull request closed without merging**
